File: DSA/src/py/copy_list_with_random_pointer.py
from typing import Optional

class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        copyList = dict()
        copyList[None] = None
        curr = head
        while curr:
            copyList[curr] = Node(curr.val)
            curr = curr.next

        copyCurr, curr = copyList[head], head
        while curr:
            copyCurr.next = copyList[curr.next]
            copyCurr.random = copyList[curr.random]
            curr = curr.next
            copyCurr = copyCurr.next
        return copyList[head]

File: docs/DSA/src/go/copy_list_with_random_pointer.go
package main

type RandomListNode struct {
	Val    int
	Next   *RandomListNode
	Random *RandomListNode
}

func copyRandomList(head *RandomListNode) *RandomListNode {
	clone := make(map[*RandomListNode]*RandomListNode)
	curr := head
	for curr != nil {
		clone[curr] = &RandomListNode{Val: curr.Val}
		curr = curr.Next
	}

	curr = head
	cloneCurr, _ := clone[curr]
	for curr != nil {
		cloneCurr.Next, _ = clone[curr.Next]
		cloneCurr.Random, _ = clone[curr.Random]
		curr = curr.Next
		cloneCurr = cloneCurr.Next
	}
	res, _ := clone[head]
	return res
}
